### PR TITLE
Added a UserWarning when using torch.{std,var,std_mean,std_var} with dof<=0

### LINKED ISSUE #109696: Standardize `UserWarnings` for `torch.cov`, `torch.{std,var}` and `torch.{std,var}_mean`
### 🐛 Describe the bug

Follow-up to #109186.

The following stats functions

- `torch.var`
- `torch.var_mean`
- `torch.std`
- `torch.std_mean`
- `torch.cov`

are inconsistent in their behavior when `correction >= N`. Only `torch.cov` throws a `UserWarning`:

```
>>> torch.var(torch.randn(3), correction=3)
tensor(inf)
>>> torch.std(torch.randn(3), correction=3)
tensor(inf)
>>> torch.var_mean(torch.randn(3), correction=3)
(tensor(inf), tensor(-0.4473))
>>> torch.std_mean(torch.randn(3), correction=3)
(tensor(inf), tensor(0.7958))
>>> torch.cov(torch.randn(2, 3), correction=3)
<stdin>:1: UserWarning: cov(): degrees of freedom is <= 0 (Triggered internally at ../aten/src/ATen/native/Correlation.cpp:117.)
tensor([[inf, -inf],
        [-inf, inf]])
```

As `correction >= N` would lead to an effective `dof=0` (and therefore `inf` values), we should always throw a `UserWarning`.

### Versions

Versions of relevant libraries:
[pip3] numpy==1.24.1
[pip3] pytorch-triton==2.1.0+6e4932cda8
[pip3] torch==2.2.0.dev20230919+cu118
[pip3] torchaudio==2.2.0.dev20230910+rocm5.6
[pip3] torchvision==0.17.0.dev20230910+rocm5.6
[pip3] triton==2.0.0

cc @albanD

### PR DESCRIPTION
**Pull request by tringwald**
Fixes #109696.

This PR adds a `UserWarning` when calling 
- `torch.var`
- `torch.var_mean`
- `torch.std`
- `torch.std_mean`

with an effective `dof<=0`. Until now, only `torch.cov` warned about this. The code also handles edge cases, such as `torch.empty`
```
>>> import torch; torch.std_mean(torch.empty(0), correction=0)
<stdin>:1: UserWarning: std_mean(): degrees of freedom is <= 0 (Triggered internally at /app/aten/src/ATen/native/ReduceOps.cpp:1671.)
(tensor(nan), tensor(nan))
```

multi-dim reductions

```
>>> import torch; torch.std_mean(torch.empty(10, 30, 20, 50), correction=600, dim=(1, 2))
<stdin>:1: UserWarning: std_mean(): degrees of freedom is <= 0 (Triggered internally at /app/aten/src/ATen/native/ReduceOps.cpp:1671.)
[... snip ...]
```

and a negative `correction`.

```
>>> import torch; torch.std_mean(torch.randn(0), correction=-5)
(tensor(nan), tensor(nan))
```

### CODE REVIEW

**Review by albanD — COMMENTED** [review]
Will let @soulitzer review this one

File: aten/src/ATen/native/ReduceOps.cpp [hunk 1]
```diff
@@ -1662,6 +1662,17 @@ static double std_var_all_cpu(const Tensor& self, double correction, bool take_s
   return result;
 }
 
+namespace {
+  static inline void warn_invalid_degrees_of_freedom(const char* fname, const TensorIterator iter, double correction) {
```
Comment (soulitzer) [inline aten/src/ATen/native/ReduceOps.cpp:1666]: nit: no need for static if we're already in anonymous namespace

File: aten/src/ATen/native/ReduceOps.cpp [hunk 2]
```diff
@@ -1662,6 +1662,17 @@ static double std_var_all_cpu(const Tensor& self, double correction, bool take_s
...
+    int64_t reducing_over_num_elements = iter.num_output_elements() == 0 ? 0 : iter.numel() / iter.num_output_elements();
+    if (reducing_over_num_elements - correction <= 0) {
+      std::ostringstream ss;
+      ss << fname << "(): degrees of freedom is <= 0";
+      TORCH_WARN(ss.str());
```
Comment (soulitzer) [inline aten/src/ATen/native/ReduceOps.cpp:1671]: We can do this in one line: `TORCH_WARN(fname, "(): degrees of freedom is <= 0");`

File: aten/src/ATen/native/ReduceOps.cpp (hunk 1 above)
Comment (soulitzer) [reply aten/src/ATen/native/ReduceOps.cpp:1666]: Can we pass in iter as a const reference instead to avoid an unnecessary copy?

**Review by soulitzer — COMMENTED** [review]
Thanks!
Could we add a small test in `test/test_reductions.py`?

```python
 with warnings.catch_warnings(record=True) as w:
        # stuff

self.assertIn('blah', str(w[0].message))
```

File: aten/src/ATen/native/ReduceOps.cpp [hunk 3]
```diff
@@ -1662,6 +1662,17 @@ static double std_var_all_cpu(const Tensor& self, double correction, bool take_s
   return result;
 }
 
+namespace {
+  static inline void warn_invalid_degrees_of_freedom(const char* fname, const TensorIterator iter, double correction) {
+    int64_t reducing_over_num_elements = iter.num_output_elements() == 0 ? 0 : iter.numel() / iter.num_output_elements();
+    if (reducing_over_num_elements - correction <= 0) {
+      std::ostringstream ss;
+      ss << fname << "(): degrees of freedom is <= 0";
```
Comment (soulitzer) [inline aten/src/ATen/native/ReduceOps.cpp:1670]: Should we include in the error message a way how the user can correct the problem?

Comment (tringwald) [reply aten/src/ATen/native/ReduceOps.cpp:1670]: Right now, the warning text is the same as in `torch.cov`, but we can of course add something like: "correction should be less than the number of elements used in the reduction operation."

Comment (soulitzer) [reply aten/src/ATen/native/ReduceOps.cpp:1670]: That would be great. Maybe something like "correction should be strictly less than the reduction factor (input numel divided by output numel)".

Comment (tringwald) [reply aten/src/ATen/native/ReduceOps.cpp:1670]: That would work for mean/std/var, but it's a bit more tricky for cov, because the reduction operations are implicitly used during the cov calculation.

```python3
>>> x = torch.randn(5, 10)
>>> x.numel()
50
>>> torch.cov(x).numel()
25
```

Correction should be fine up to `N-1 = 9` here.

Comment (tringwald) [reply aten/src/ATen/native/ReduceOps.cpp]: I used your proposed message for the normal reduction ops and provided a more specific message for `torch.cov` w.r.t. the number of observations.

**Review by soulitzer — APPROVED** [review]
Looks great, thanks for the update!
